src/librbd/cache/pwl/InitRequest.cc

```diff
@@ -125,6 +125,13 @@ void InitRequest<I>::init_image_cache() {
   Context *ctx = create_context_callback<
     klass, &klass::handle_init_image_cache>(this);
   m_image_cache->init(ctx);
+
+  if (m_error_result < 0) {
```

Comment:
I am not sure I like the approach much. Your assumption is that in the case of init failure, `handle_init_image_cache` will be always called in this context before you are checking `m_error_result`.

Although it may be correct it is rather fragile (may change in the future) and the code written this way is very confusing -- it took some time to me to understand it.

Our usual approach to resolve cases like this, if it is not a fast io path, is to use `create_async_context_callback` instead of `create_context_callback` to make sure the handle function is never called in the same context.

Another approach would be to do deletion of `m_image_cache` in `m_onfinish` context, which as I see is already wrapped with `create_async_context_callback`, though I would vote for the first approach as it looks to me a little clearer.